Mathlib/NumberTheory/ModularForms/Basic.lean

```diff
@@ -397,3 +399,125 @@ instance (priority := 99) [CuspFormClass F Γ k] : ModularFormClass F Γ k where
   bdd_at_infty _ _ := (CuspFormClass.zero_at_infty _ _).boundedAtFilter
 
 end CuspForm
+
+namespace ModularForm
+
+section GradedRing
+
+/--cast for modular forms, which is useful for removing `heq`'s.-/
+def mcast {a b : ℤ} {Γ : Subgroup SL(2, ℤ)} (h : a = b) (f : ModularForm Γ a) : ModularForm Γ b
+    where
+  toFun := (f : ℍ → ℂ)
+  slash_action_eq' := by intro A; have := f.slash_action_eq' A; convert this; exact h.symm
+  holo' := f.holo'
+  bdd_at_infty' := by intro A; convert f.bdd_at_infty' A <;> exact h.symm
+
+theorem type_eq {a b : ℤ} (Γ : Subgroup SL(2, ℤ)) (h : a = b) :
+    ModularForm Γ a = ModularForm Γ b := by
+  induction h
+  rfl
+
+theorem cast_eq_mcast {a b : ℤ} {Γ : Subgroup SL(2, ℤ)} (h : a = b) (f : ModularForm Γ a) :
+    cast (type_eq Γ h) f = mcast h f := by
+  induction h
+  ext1
+  rfl
+
+theorem hEq_one_mul (k : ℤ) {Γ : Subgroup SL(2, ℤ)} (f : ModularForm Γ k) :
+    HEq ((1 : ModularForm Γ 0).mul f) f := by
+  apply heq_of_cast_eq (type_eq Γ (zero_add k).symm).symm
+  funext
+  rw [cast_eq_mcast, mcast]
+  simp only [mul_coe, one_coe_eq_one, one_mul]
+  ext1
+  rfl
+  simp only [zero_add]
```

Comment:
You've got a few repetitive proofs here that could be avoided like this:
```suggestion
theorem hEq_of_mcast_eq {Γ : Subgroup SL(2, ℤ)} (h : a = b)
    {f : ModularForm Γ a} {g : ModularForm Γ b} (H : ∀ x, f x = g x) : HEq f g := by
  apply heq_of_cast_eq (type_eq Γ h)
  rw [cast_eq_mcast h]
  ext1
  simp [H]

theorem hEq_one_mul (k : ℤ) {Γ : Subgroup SL(2, ℤ)} (f : ModularForm Γ k) :
    HEq ((1 : ModularForm Γ 0).mul f) f :=
  hEq_of_mcast_eq (zero_add k) fun _ => by simp
```

Reply:
I think this is basically the same as https://github.com/leanprover-community/mathlib4/pull/9164#discussion_r1432765465, but my version avoids the HEq entirely.